### PR TITLE
[AArch64] Unroll some loops with early-continues on Apple Silicon.

### PR DESCRIPTION
**Pull request by fhahn**
Try to runtime-unroll loops with early-continues depending on loop-varying
loads; this helps with branch-prediction for the early-continues and can
significantly improve performance for such loops

Builds on top of https://github.com/llvm/llvm-project/pull/118317

### CODE REVIEW

**Comment by fhahn** [comment]
Rebased after landing https://github.com/llvm/llvm-project/pull/118317.

Ping :)

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 1]
```diff
@@ -4068,51 +4068,86 @@ getAppleRuntimeUnrollPreferences(Loop *L, ScalarEvolution &SE,
...
+    return;
   }
 
-  if (BestUC == 1 || none_of(Stores, [&LoadedValues](StoreInst *SI) {
-        return LoadedValues.contains(SI->getOperand(0));
-      }))
+  // Try to runtime-unroll loops with early-continues depending on loop-varying
+  // loads; this helps with branch-prediction for the early-continues.
+  auto *Term = dyn_cast<BranchInst>(Header->getTerminator());
+  auto *Latch = L->getLoopLatch();
```
Comment (jroelofs) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:4124]: is `Latch` guaranteed to be non-null? I see other places where there are tests for `if (!L->getLoopLatch())`, and yet more where there are `assert(L->getLoopLatch());`

Comment (jroelofs) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:4124]: oh, this whole thing is guarded with `if (Header == L->getLoopLatch())`

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 2]
```diff
@@ -4068,51 +4068,86 @@ getAppleRuntimeUnrollPreferences(Loop *L, ScalarEvolution &SE,
...
+      return true;
+
+    return any_of(I->operands(), [&](Value *V) {
+      auto *I = dyn_cast<Instruction>(V);
+      return I && DependsOnLoopLoad(I, Depth + 1);
+    });
+  };
+  CmpInst::Predicate Pred;
+  Instruction *I;
+  if (match(Term, m_Br(m_ICmp(Pred, m_Instruction(I), m_Value()), m_Value(),
```
Comment (jroelofs) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:4146]: what about a `DependsOnLoopLoad(I, 0)` via the other `ICmp` operand, would that be canonicalized away, or should we care about that here?

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp [hunk 3]
```diff
@@ -4068,51 +4068,86 @@ getAppleRuntimeUnrollPreferences(Loop *L, ScalarEvolution &SE,
...
     return;
 
-  UP.Runtime = true;
-  UP.DefaultUnrollRuntimeCount = BestUC;
+  std::function<bool(Instruction *, unsigned)> DependsOnLoopLoad =
+      [&](Instruction *I, unsigned Depth) -> bool {
+    if (isa<PHINode>(I) || L->isLoopInvariant(I) || Depth > 8)
+      return false;
+
+    if (auto *LI = dyn_cast<LoadInst>(I))
```
Comment (jroelofs) [inline llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:4136]: `isa`, since the result isn't used.

**Review by jroelofs — APPROVED** [review]
couple of nits/questions, but otherwise LGTM

File: llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:4163]: For now I think most cases should be caught due to canonicalization. I'll look into extending it as follow up

**Pull request merged**